base/threadingconstructs.jl

```diff
@@ -110,15 +111,73 @@ A barrier is placed at the end of the loop which waits for all tasks to finish
 execution.
 
 The `schedule` argument can be used to request a particular scheduling policy.
-The only currently supported value is `:static`, which creates one task per thread
-and divides the iterations equally among them. Specifying `:static` is an error
-if used from inside another `@threads` loop or from a thread other than 1.
+
+Except for `:static` scheduling, how the iterations are assigned to tasks, and how the tasks
+are assigned to the worker threads is undefined. The exact assignments can be different
+for each execution. The scheduling option is a hint. The loop body code (including any code
+transitively called from it) must not make assumptions about the distribution of iterations
+to tasks or the worker thread in which they are executed. The loop body for each iteration
+must be able to make forward progress independent of other iterations and be free from data
+races. As such, synchronizations across iterations may deadlock.
+
+For example, the above conditions imply that:
+
+- The lock taken in an iteration *must* be released within the same iteration.
+- Communicating between iterations using blocking primitives like `Channel`s is incorrect.
+- Write only to locations not shared across iterations (unless a lock or atomic operation is used).
+
+
+Schedule options are:
+- `:static` creates one task per thread and divides the iterations equally among
+            them, assigning each task specifically to each thread.
+            Specifying `:static` is an error if used from inside another `@threads` loop
+            or from a thread other than 1.
+- `:dynamic` will schedule iterations dynamically to available worker threads,
+            assuming that the workload for each iteration is uniform.
+
+Without the scheduler argument, the exact scheduling is unspecified; i.e. it may be
+different across Julia releases. Currently, the behavior is dependent on the calling thread.
+When called from thread 1 the default is `:static`. When called from other threads the loop
+will be executed without threading.
 
 The default schedule (used when no `schedule` argument is present) is subject to change.
 
+For example, an illustration of the different scheduling strategies where `busywait`
+is a non-yielding timed loop that runs for a number of seconds.
+
+```julia-repl
+julia> function busywait(seconds)
+            tstart = Base.time()
+            while Base.time() - tstart < seconds
```

Comment:
How about making it work even through the leap second and if the user resets the system time? :laughing: 

```suggestion
            tstart = time_ns()
            while (time_ns() - tstart) / 1e9 < seconds
```

Technically, we also need `GC.safepoint()` in the `while` loop. But maybe it's better to keep the example simple.